cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,335 @@
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long checkJobQueue() {
+      boolean emptyFetch = false;
+      try {
+        emptyFetch = Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            return checkJobConstraints(context);
+          }
+        });
+        failureCount = 0;
+      } catch (Exception e) {
+        LOG.warn("Failed to check Job constraints. Will retry in next run", e);
+        failureCount++;
+      }
+
+      try {
+        Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            // run any ready jobs
+            return runReadyJobs(context);
+          }
+        });
+      } catch (Exception e) {
+        LOG.warn("Failed to launch programs. Will retry in next run", e);
+        failureCount++;
+      }
+
+      // If there is any failure, delay the next fetch based on the strategy
+      if (failureCount > 0) {
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+
+      // Sleep for 2 seconds if there's no jobs in the queue
+      return emptyFetch ? 2000L : 0L;
+    }
+
+    private boolean checkJobConstraints(DatasetContext context) throws Exception {
+      boolean emptyScan = true;
+
+      JobQueueDataset jobQueue = getJobQueue(context);
+      CloseableIterator<Job> jobQueueIter;
+      if (lastConsumed == null) {
+        jobQueueIter = jobQueue.getJobs(partition);
+      } else {
+        jobQueueIter = jobQueue.getJobs(partition, lastConsumed);
+      }
+
+      try {
+        Stopwatch stopWatch = new Stopwatch().start();
+        // limit the batches of the scan to 1000ms
+        while (isRunning() && stopWatch.elapsedMillis() < 1000) {
+          if (!jobQueueIter.hasNext()) {
+            jobQueueIter.close();
+            lastConsumed = null;
+            return emptyScan;
+          }
+          Job job = jobQueueIter.next();
+          lastConsumed = job;
+          emptyScan = false;
+          checkAndUpdateJob(jobQueue, job);
+        }
+      } finally {
+        jobQueueIter.close();
+      }
+      return emptyScan;
+    }
+
+    private void checkAndUpdateJob(JobQueueDataset jobQueue, Job job) {
+      switch (job.getState()) {
+        case PENDING_TRIGGER:
+          if (isTriggerSatisfied(job)) {
+            jobQueue.updateState(job, Job.State.PENDING_CONSTRAINT);
+          }
+        // fallthru to check for constraint satisfaction after trigger satisfaction
+        case PENDING_CONSTRAINT:
+        case PENDING_LAUNCH:
+          if (constraintsSatisfied(job)) {
+            jobQueue.updateState(job, Job.State.PENDING_LAUNCH);
+            readyJobs.add(job);
+          }
```

Comment:
If its in PENDING_LAUNCH, its safer to leave it in PENDING_LAUNCH, since there may have been another Job created after it. The semantics are clearer this way.